config/module.go

```diff
@@ -233,72 +245,212 @@ const FirstRunSuccessSuffix = ".first_run_succeeded"
-// a marker file indicating that the setup phase has run successfully.
-func (m Module) EvaluateFirstRunPath(packagesDir string, logger logging.Logger) (
-	string,
-	func() error,
-	error,
-) {
-	noop := func() error { return nil }
-	unpackedModDir, err := m.exeDir(packagesDir)
+func (m Module) EvaluateFirstRunPath(unpackedModDir string, logger logging.Logger) (string, error) {
+	// note: we don't look at internal meta.json in local non-tarball case because user has explicitly requested a binary.
+	localNonTarball := m.Type == ModuleTypeLocal && !m.NeedsSyntheticPackage()
+	if !localNonTarball {
+		// this is case 1, meta.json in exe folder.
+		metaPath, err := utils.SafeJoinDir(unpackedModDir, "meta.json")
+		if err != nil {
+			return "", err
+		}
+		_, err = os.Stat(metaPath)
+		if err == nil {
+			// this is case 1, meta.json in exe dir
+			meta, err := parseJSONFile[JSONManifest](metaPath)
+			if err != nil {
+				return "", err
+			}
+			firstRun, err := utils.SafeJoinDir(unpackedModDir, meta.FirstRun)
+			if err != nil {
+				return "", err
+			}
+			firstRunPath, err := filepath.Abs(firstRun)
+			return firstRunPath, err
+		}
+	}
+	if m.NeedsSyntheticPackage() {
+		// this is case 2, side-by-side
+		// TODO(RSDK-7848): remove this case once java sdk supports internal meta.json.
+		metaPath, err := utils.SafeJoinDir(filepath.Dir(m.ExePath), "meta.json")
+		if err != nil {
+			return "", err
+		}
+		meta, err := parseJSONFile[JSONManifest](metaPath)
+		if err != nil {
+			// note: this error deprecates the side-by-side case because the side-by-side case is deprecated.
+			return "", errors.Wrapf(err, "couldn't find meta.json inside tarball %s (or next to it)", m.ExePath)
+		}
+		firstRun, err := utils.SafeJoinDir(unpackedModDir, meta.FirstRun)
+		if err != nil {
+			return "", err
+		}
+		firstRunPath, err := filepath.Abs(firstRun)
+		return firstRunPath, err
+	}
+	return "", errors.New("no first run script")
+}
+
+// FirstRun executes a module-specific setup script.
+func (m *Module) FirstRun(
+	ctx context.Context,
+	localPackagesDir,
+	dataDir string,
+	env map[string]string,
+	logger logging.Logger,
+) error {
+	logger = logger.Sublogger("first_run").WithFields("module", m.Name)
+
+	unpackedModDir, err := m.exeDir(localPackagesDir)
 	if err != nil {
-		return "", noop, err
+		return err
 	}
 
+	// check if FirstRun already ran successfully for this module version by
+	// checking if a success marker file exists on disk.
 	firstRunSuccessPath := unpackedModDir + FirstRunSuccessSuffix
 	if _, err := os.Stat(firstRunSuccessPath); !errors.Is(err, os.ErrNotExist) {
 		logger.Info("first run already ran")
-		return "", noop, errors.New("first run already ran")
+		return nil
+	}
+
+	// Load the module's meta.json. If it doesn't exist or fails to load, debug log and exit quietly.
+	meta, err := m.getJSONManifest(unpackedModDir)
+	if err != nil {
+		logger.Debugw("failed to load meta.json, skipping setup phase", "error", err)
+		return nil
+	}
```

Comment:
this error can happen if:

1. meta.json is missing
2. meta.json cannot be parsed

a `DEBUG` log for 1. feels appropriate. I can see an argument for having a higher log level for 2. This would require a bit more refactoring to achieve so curious how other feel about this.

Reply:
[opt-nit] Is there any way to differentiate between 1 and 2 and `Debug` on first and `Warn` on second?

Reply:
I'll try